### ISSUE TITLE
[REQUEST] coloring of the output

### ISSUE THREAD

**Issue by DBJDBJ** [compiler-explorer/compiler-explorer#2405]
https://github.com/sheredom/ubench.h/issues/11#issue-796689452

**The solution I would like**
A way to talk to the JavaScript/CSS in order to achieve "VT100 like" coloring. 

"Godbolt Terminal Escape Codes (GTEC)", perhaps :)

The whole UI is JavaScript. I might think that might be doable? 

```cpp
// using Godbolt Terminal Escape Codes (GTEC)
// GTEC_FG_RED == Godbolt Terminal Foreground Red
printf( GTEC_FG_RED "Error?!" GTEC_FG_RESET ) ;
```

The mechanism might be the same of very similar to javascript driven source code syntax coloring.

**Alternatives considered**
I can't think of any.

**Comment by apmorton** [comment]
We already support standard color escape sequences.

ubench is simply not outputting colors because it doesn't detect stdout as being a tty:
https://github.com/sheredom/ubench.h/blob/master/ubench.h#L207

I believe emulating a true tty has been suggested here before, but I don't know where we land on that yet.

In the meantime, to get color output with ubench you would need an upstream change in ubench to override the tty detection

**Comment by DBJDBJ** [comment]
@apmorton  many thanks for the quick reply, with all the respect I know that and this is why I am "here" :)

I might claim VT100 escape code ("tty" et all) emulation is one very bold task, which implementation very likely will not be that much used. 

But, "Simple" syntax colouring of the Godbolt IDE output windows is perhaps more likely to be implemented.  I am sure "someone somewhere" has already implemented in javascript an "terminal" with "colors". "We" just need "hooks" we can use from C/C++ macros.

ps: I  might make a quick ubench.h fork that is not using tty check.  But how  is that going to work? The Godbolt output is not a terminal but javascript emulation, right? Or I might be completely wrong :)

**Comment by partouf** [comment]
We pipe the stdout and stderr streams to a js buffer and return those to the UI. If it includes coloring, it is done through escape codes and those are retained in the streams.
#2315 mentions a basic example that shows which escape codes are supported.

**Comment by DBJDBJ** [comment]
See the line 10 and 11 .. using my fork

https://godbolt.org/z/zPncs1

And that does not bring coloring since Godbolt ide output is not a TTY, but a JavaScript emulation; I assume.

javascript syntax  "coloring concept" might work here .. yes that requires some non-trivial development, I admit. javascript will have to filter the output, look for "special symbols" and replace them with css text coloring. 

Javascript: 

if text contains "{{GTEC_FG_RED}}" replace it with `<span class="open_fg_red" >`
if text contains "{{GTEC_FG_RESET}}" replace it with `</span>`

C/C++

#define GTEC_FG_RED "{{GTEC_FG_RED}}"
#define GTEC_FG_RESET "{{GTEC_FG_RESET}}"

That should work as a concept.

**Comment by partouf** [comment]
https://godbolt.org/z/cs4GYz works

**Comment by DBJDBJ** [comment]
@partouf  just seen your reply let me try thanks

**Comment by DBJDBJ** [comment]
> https://godbolt.org/z/cs4GYz works

And it worketh ...

You have added Line 8 

![image](https://user-images.githubusercontent.com/145155/106273297-b46f1e00-6232-11eb-950a-4fe58735dab0.png)

Can you please elaborate?

**Comment by partouf** [comment]
This is the difference:
![image](https://user-images.githubusercontent.com/442630/106273618-40814580-6233-11eb-95ee-9a8441475ea1.png)

The main problem being the original header being included before your own, which makes your own fork not actually work.

The #include is inlined and because the headers are guarded like this:
```
#ifndef SHEREDOM_UBENCH_H_INCLUDED
#define SHEREDOM_UBENCH_H_INCLUDED
```
The compiler won't actually parse your fork.

The #define FALSE is just a bugfix.

**Comment by DBJDBJ** [comment]
> The #define FALSE is just a bugfix.

Of which bug? Thanks :)

**Comment by DBJDBJ** [comment]
Just ignore me :)

![image](https://user-images.githubusercontent.com/145155/106275603-665c1980-6236-11eb-9c9e-10c94aa20ad5.png)

**Comment by partouf** [comment]
> > The #define FALSE is just a bugfix.
> 
> Of which bug? Thanks :)

No idea, I know there are differences in where FALSE is defined in the Windows headers and in the Linux compiler headers, maybe you mixed up some ifdefs in your fork.

**Comment by mattgodbolt** [comment]
> javascript syntax "coloring concept" might work here .. yes that requires some non-trivial development, I admit. javascript will have to filter the output, look for "special symbols" and replace them with css text coloring.
> 
> Javascript:
> 
> if text contains "{{GTEC_FG_RED}}" replace it with `<span class="open_fg_red" >`
> if text contains "{{GTEC_FG_RESET}}" replace it with `</span>`

[snip]

I'm going to close this issue - I don't think us inventing our own colourization output is a good idea when there's already a supported standard for this. Whether the application detects the output is a TTY and has to be overridden, or the application has to output non-standard `{{GTEC_FG_RED}}` things, either way the app needs to change, and it might as well use ANSI codes :)

**Comment by DBJDBJ** [comment]
@mattgodbolt as far as I can see this works. No changes were necessary. See above.
